### PR TITLE
removes is_exited:Arc<AtomicBool> from PohRecorder

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
In order to avoid banking stage threads being blocked
indefinitely at exit time
```rust
is_exited: Arc<AtomicBool>
```
was added in https://github.com/solana-labs/solana/pull/16503 to `PohRecorder`.

#### Summary of Changes

To simplify this code for follow up changes, this commit removes
`is_exited` from `PohRecorder` and instead sends
```rust
Err(PohRecorderError::MaxHeightReached)
```
to waiting threads and then drops the receiver end of the channel
immediately.
Running the problematic tests repeatedly shows that this is sufficient
to unblock threads at exit time.

### CODE REVIEW

File: poh/src/poh_service.rs [hunk 1]
```diff
@@ -134,17 +132,23 @@ impl PohService {
...
                     );
                 }
-                poh_exit_.store(true, Ordering::Relaxed);
+                poh_exit.store(true, Ordering::Relaxed);
+                // Drain record_receiver channel and unblock remaining threads
+                // waiting on the receiver end of the nested channels.
+                for record in record_receiver.try_iter() {
+                    let _ = record.sender.send(Err(PohRecorderError::MaxHeightReached));
+                }
+                drop(record_receiver);
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply poh/src/poh_service.rs:151]: How would you then call `.try_iter()` then?

[the remark this answers is not in the document]
Comment (behzadnouri) [reply poh/src/poh_service.rs:151]: >  to avoid new items from arriving after the initial try_iter()

There is no "initial" `try_iter`.
The for loop keeps running as long as there are items on the channel even if those items were pushed _after the for loop initially started_.

Also note that there is a
```rust
poh_exit.store(true, Ordering::Relaxed);
```
just before the for loop starts which should stop other validator processes.

`try_iter` returns an iterator which keeps inspecting if the channel is empty or not. As soon as the channel is empty, the iterator returns `None`, the for loop ends, and we drop the receiver side of the channel, which disconnects it from sender and stops sender from sending new items.

_Theoretically_ there is still some minuscule chance that new items are pushed between the time the channel becomes empty and `drop` is invoked. But having run tests for 1000s times, I have not seen that happen once;

[the remark this answers is not in the document]
Comment (carllin) [reply poh/src/poh_service.rs:151]: > try_iter returns an iterator which keeps inspecting if the channel is empty or not. As soon as the channel is empty, the iterator returns None, the for loop ends, and we drop the receiver side of the channel, which disconnects it from sender and stops sender from sending new items.

Yeah these subtle edge cases usually have a way of showing up in tests and other odd behavior. I've also seen times where we run the tests a bunch locally and it seems fine, but then shows up in CI and other people's machines.

I think I would prefer correctness for these subtle edge cases, even if it means introducing more complexity as these are hard to go back and debug/build on top of.
